### PR TITLE
RAC-4375-Update submodules of RackHD, for every sprint release

### PR DESCRIPTION
**Pull request by kellylu2sym**
Update submodule
background : 
In RackHD repository , there are several submodules , we should update the submodule automatically  to make sure the submodules related to latest tested commit. 
Details : 
we add steps after "bump version" to update the submodules version. if the tests are failed ,it will be reverted . 
test passed : http://10.62.59.175:8080/job/springReleaseKelly/169/

reviewer : @panpan0000  @changev @PengTian0

### CODE REVIEW

File: build-release-tools/application/update_submodule.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+            manifest.validate_manifest()
+    except KeyError as error:
+            print "Failed to create a Manifest instance for the manifest file {0} \nERROR:\n{1}"\
+                  .format(args.manifest, error.message)
+            sys.exit(1)
+
+
+    if args.publish:
+        if args.git_credential:
+            repo_operator = RepoOperator(args.git_credential)
```
Comment (panpan0000) [inline build-release-tools/application/update_submodule.py]: why inside args.publish ?

Comment (panpan0000) [inline build-release-tools/application/update_submodule.py:91]: if args.publish == false.
then 
```
repo_operator.submodule_init(repo_dir)
repo_operator.submodule_update(repo_dir) 
```
will all fail

I think you can 
```
if args.publish:
      if args.git_credential:
            repo_operator = RepoOperator(args.git_credential)
else
      repo_operator = RepoOperator()
```

File: build-release-tools/lib/RepositoryOperator.py [hunk 2]
```diff
@@ -541,3 +541,161 @@ def push_repo_changes(self, repo_dir, commit_message, push_all=False):
...
+        Init submodule of repository 
+        :param repo_dir: the directory of the repository
+        :return: None
+        """
+       
+        if repo_dir is None or not os.path.isdir(repo_dir):
+            raise RuntimeError("The repository directory {0} is not specified. or its format is wrong".format(repo_dir))
+
+        return_code, output, error = self.git.run(['submodule', 'init'], directory=repo_dir)
+	print "=========== debug in submoduel_init"
```
Comment (panpan0000) [inline build-release-tools/lib/RepositoryOperator.py]: can we remove it ?

**Comment by panpan0000** [comment]
I'm ok in general, with some small comments.

@PengTian0  to review as well

File: jobs/SprintRelease/Jenkinsfile [hunk 3]
```diff
@@ -42,6 +43,8 @@ node{
                 if(JUMP_VERSION == "true"){
                     load("jobs/SprintRelease/bump_version.groovy")
                 }
+                
+                  submodule.submodule("update");
```
Comment (PengTian0) [inline jobs/SprintRelease/Jenkinsfile]: If the submodule is updated, the RackHD will have a new commit, right?
If so, the commit of RackHD in manifest is not the latest commit.
The tested RackHD is still the version without updating submodules.
Why not update submodules after unit test and function test?

File: jobs/SprintRelease/update_submodule.groovy [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+                        timeout(5){
+                            sh '''#!/bin/bash -ex
+                            # Update the submodule according to the manifest file.
+                            ./build-config/build-release-tools/HWIMO-BUILD build-config/build-release-tools/application/${OPERATION}_submodule.py \
+                            --build-dir b \
+                            --manifest build-config/rackhd-manifest \
+                            --publish \
+                            --version ${tag_name}\
+                            --git-credential https://github.com/kellylu2sym,JENKINSRHD_GITHUB_CREDS
+                            '''
```
Comment (PengTian0) [inline jobs/SprintRelease/update_submodule.groovy]: Please update it to production url

File: build-release-tools/application/revert_submodule.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+                if len(submodules_list)==0:
+                    continue;
+                #git pull is used to get the lastest code to make sure push successfully
+                test_info = repo_operator.git_pull(repo_dir)
+                update_commit_message="update submodule for new commit {0}".format(args.version)
+                commit_info = repo_operator.get_commit_of_update_submodule(repo_dir,update_commit_message)
+                commit_id = get_previous_id(commit_info)
+                if commit_id != None:
+                        repo_operator.revert_to_commit(repo_dir,commit_id)
+                if args.publish:
```
Comment (PengTian0) [inline build-release-tools/application/revert_submodule.py]: Only when this line ```repo_operator.revert_to_commit(repo_dir,commit_id)``` is executed, the change needs to be published ,right?

File: build-release-tools/application/revert_submodule.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        print args.build_dir
+        for filename in os.listdir(args.build_dir):
+            try:
+                repo_dir = os.path.join(args.build_dir, filename)
+                repo_operator.submodule_init(repo_dir)
+                repo_operator.submodule_update(repo_dir)
+
+                submodules_list = repo_operator.get_current_submodule(repo_dir)
+                if len(submodules_list)==0:
+                    continue;
```
Comment (PengTian0) [inline build-release-tools/application/revert_submodule.py]: Could you please refactor line 110~116 to a method like: subModulesExist?

File: build-release-tools/application/revert_submodule.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+                if args.publish:
+                    print "start to publish  revert update submodule in {0}".format(repo_dir)
+                    commit_message = "revert update submodule for new commit {}".format(args.version)
+                    repo_operator.push_repo_changes(repo_dir, commit_message)
+            except Exception,e:
+                print "Failed to revert update submodule of {0} due to {1}".format(filename, e)
+                sys.exit(1)
+    else:
+        print "The argument build-dir must be a directory"
+        sys.exit(1)
```
Comment (PengTian0) [inline build-release-tools/application/revert_submodule.py:141]: Could you please refactor line 117~133 to a method like: revertCommitForUpdateSubModules?

File: jobs/SprintRelease/update_submodule.groovy [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+                        }
+                    }
+                }
+            }
+            stage("Update Manifest"){
+                retry(retry_times){
+                    timeout(3){
+                      //  sh './build-config/jobs/SprintRelease/update_manifest.sh'
```
Comment (PengTian0) [inline jobs/SprintRelease/update_submodule.groovy]: Is used for debug?

File: jobs/SprintRelease/update_submodule.groovy [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+                    }
+                }
+                // inject properties file as environment variables
+                if(fileExists ('downstream_file')) {
+                    //def props = readProperties file: 'downstream_file';
+                    if(props['MANIFEST_FILE_URL']) {
+                       // env.MANIFEST_FILE_URL = "${props.MANIFEST_FILE_URL}";
```
Comment (PengTian0) [inline jobs/SprintRelease/update_submodule.groovy]: The same with above

File: jobs/SprintRelease/update_submodule.groovy [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+            }
+            stage("Update Manifest"){
+                retry(retry_times){
+                    timeout(3){
+                      //  sh './build-config/jobs/SprintRelease/update_manifest.sh'
+                    }
+                }
+                // inject properties file as environment variables
+                if(fileExists ('downstream_file')) {
+                    //def props = readProperties file: 'downstream_file';
```
Comment (PengTian0) [inline jobs/SprintRelease/update_submodule.groovy]: The same with above

File: jobs/SprintRelease/Jenkinsfile [hunk 11]
```diff
@@ -68,7 +71,13 @@ node{
                         // Run function test
                         source_based_test.runTests(function_test)
                     }
-                } finally{
+                }catch(Exception e){
+
+                     submodule.submodule("revert");   
+
```
Comment (PengTian0) [inline jobs/SprintRelease/Jenkinsfile]: Please re-throw exception by:
```error("[ERROR]...")```

**Comment by PengTian0** [comment]
👍

**Review by panpan0000 — APPROVED** [review]

**Review by iceiilin — APPROVED** [review]

**Pull request merged**